dali/pipeline/operators/decoder/decoder_cache_blob.h

```diff
@@ -0,0 +1,201 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DECODER_DECODER_CACHE_BLOB_H_
+#define DALI_PIPELINE_OPERATORS_DECODER_DECODER_CACHE_BLOB_H_
+
+#include <string>
+#include <mutex>
+#include <unordered_map>
+#include "dali/pipeline/data/tensor_list.h"  // needed for Dims
+#include "dali/kernels/span.h"
+#include "dali/error_handling.h"
+
+namespace dali {
+
+class DecoderCacheBlob {
+ public:
+    using ImageKey = std::string;
+
+    inline DecoderCacheBlob(std::size_t cache_size,
+                            std::size_t image_size_threshold,
+                            bool stats_enabled = true )
+        : cache_size_(cache_size)
+        , image_size_threshold_(image_size_threshold)
+        , stats_enabled_(stats_enabled) {
+        DALI_ENFORCE(image_size_threshold <= cache_size_,
+            "Cache size should fit at least one image");
+    }
+
+    inline virtual ~DecoderCacheBlob() {
+        if (buffer_)
+            GPUBackend::Delete(buffer_, 0, false);
+        if (stats_enabled_ && images_seen() > 0)
+            print_stats();
+    }
+
+    DISABLE_COPY_MOVE_ASSIGN(DecoderCacheBlob);
+
+    inline bool IsCached(const ImageKey& image_key) const {
+        std::lock_guard<std::mutex> lock(mutex_);
+        return cache_.find(image_key) != cache_.end();
+    }
+
+    inline const Dims& GetShape(const ImageKey& image_key) const {
+        std::lock_guard<std::mutex> lock(mutex_);
+        const auto it = cache_.find(image_key);
+        DALI_ENFORCE(it != cache_.end(),
+            "cache entry [" + image_key + "] not found");
+        return it->second.dims;
+    }
+
+    inline void CopyData(const ImageKey& image_key,
+                         void* destination_buffer,
+                         cudaStream_t stream = 0) const {
+        std::lock_guard<std::mutex> lock(mutex_);
+        LOG_LINE << "CopyData: image_key[" << image_key << "]" << std::endl;
+        DALI_ENFORCE(!image_key.empty());
+        DALI_ENFORCE(destination_buffer != nullptr);
+        const auto it = cache_.find(image_key);
+        DALI_ENFORCE(it != cache_.end(),
+            "cache entry [" + image_key + "] not found");
+        const auto &data = it->second.data;
+        assert(data.data() < tail_);
+        assert(data.data() + data.size() <= tail_);
+        MemCopy(destination_buffer,
+                data.data(),
+                data.size(),
+                stream);
+
+        if (stats_enabled_)
+            stats_[image_key].reads++;
+    }
+
+    inline virtual void Add(const ImageKey& image_key,
+                            const uint8_t *data, std::size_t data_size,
+                            const Dims& data_shape,
+                            cudaStream_t stream = 0) {
+        if (stats_enabled_)
+            stats_[image_key].decodes++;
+
+        if (data_size < image_size_threshold_)
```

Comment:
This condition should not be enforced when the cache policy is `Largest`

Reply:
In that case image_size_threshold_ should be forced to 0. I will change that